diff -uNr grub-2.02~beta3/grub-core/boot/i386/pc/boot.S grub-2.02~beta3-aos/grub-core/boot/i386/pc/boot.S
--- a/grub-core/boot/i386/pc/boot.S	2015-09-02 22:10:53.000000000 +0200
+++ b/grub-core/boot/i386/pc/boot.S	2016-08-31 15:17:36.508808830 +0200
@@ -250,7 +250,7 @@
 	pushw	%dx
 
 	/* print a notification message on the screen */
-	MSG(notification_string)
+	/* MSG(notification_string) */
 
 	/* set %si to the disk address packet */
 	movw	$disk_address_packet, %si
diff -uNr grub-2.02~beta3/grub-core/boot/i386/pc/diskboot.S grub-2.02~beta3-aos/grub-core/boot/i386/pc/diskboot.S
--- a/grub-core/boot/i386/pc/diskboot.S	2015-09-02 22:10:53.000000000 +0200
+++ b/grub-core/boot/i386/pc/diskboot.S	2016-08-31 15:18:54.762957906 +0200
@@ -51,9 +51,9 @@
 	pushw	%dx
 
 	/* print a notification message on the screen */
-	pushw	%si
-	MSG(notification_string)
-	popw	%si
+/*	pushw	%si */
+/*	MSG(notification_string) */
+/*	popw	%si */
 
 	/* this sets up for the first run through "bootloop" */
 	movw	$LOCAL(firstlist), %di
diff -uNr grub-2.02~beta3/grub-core/kern/main.c grub-2.02~beta3-aos/grub-core/kern/main.c
--- a/grub-core/kern/main.c	2016-02-21 17:20:08.000000000 +0100
+++ b/grub-core/kern/main.c	2016-08-31 15:20:41.285893616 +0200
@@ -271,9 +271,11 @@
   grub_boot_time ("After machine init.");
 
   /* Hello.  */
+ /* 
   grub_setcolorstate (GRUB_TERM_COLOR_HIGHLIGHT);
   grub_printf ("Welcome to GRUB!\n\n");
   grub_setcolorstate (GRUB_TERM_COLOR_STANDARD);
+*/
 
   /* Init verifiers API. */
   grub_verifiers_init ();
