diff -urp a/grub-core/kern/misc.c b/grub-core/kern/misc.c
--- a/grub-core/kern/misc.c	2021-05-18 14:25:34.130000000 +0200
+++ b/grub-core/kern/misc.c	2021-05-18 14:39:45.440000000 +0200
@@ -174,6 +174,56 @@ grub_debug_enabled (const char * conditi
   return 0;
 }
 
+/* blind debug code follows */
+#include <grub/speaker.h>
+#include <grub/time.h>
+
+void
+grub_debug_play_number (unsigned long x)
+{
+  if (!grub_env_get ("debug"))
+    return;
+
+  char *tmp = grub_xasprintf ("%lu", x);
+  char *p = tmp;
+  while (*p)
+    {
+      int i;
+      for (i = 1; i <= (((*p) == '0') ? 10 : ((*p) - '0')); ++i)
+	{
+	  grub_speaker_beep_on ((p == tmp) ? 600 : 300);
+	  grub_millisleep (100);
+	  grub_speaker_beep_off ();
+	  grub_millisleep (50);
+	}
+      grub_millisleep (300);
+      p++;
+    }
+  grub_free (tmp);
+}
+
+void
+grub_debug_envvar (const char *condition, const char *fmt, ...)
+{
+  va_list args;
+
+  if (!grub_debug_enabled (condition))
+    return;
+
+  const char *debugstr;
+  va_start (args, fmt);
+  if (!is_dprintf_active)
+    debugstr = grub_xasprintf ("blinddebug/%s: %s\n", condition,
+      grub_xvasprintf (fmt, args));
+  else
+    debugstr = grub_xvasprintf (fmt, args);
+  va_end (args);
+  grub_env_set ("debug_output", grub_xasprintf ("%s%s",
+    grub_env_get ("debug_output") ? : "", debugstr));
+}
+
+int is_dprintf_active = 0;
+
 void
 grub_real_dprintf (const char *file, const int line, const char *condition,
 		   const char *fmt, ...)
@@ -182,11 +232,13 @@ grub_real_dprintf (const char *file, con
 
   if (grub_debug_enabled (condition))
     {
+      is_dprintf_active = 1;
       grub_printf ("%s:%d: ", file, line);
       va_start (args, fmt);
       grub_vprintf (fmt, args);
       va_end (args);
       grub_refresh ();
+      is_dprintf_active = 0;
     }
 }
 
diff -urp a/grub-core/kern/term.c b/grub-core/kern/term.c
--- a/grub-core/kern/term.c	2021-05-18 14:25:34.130000000 +0200
+++ b/grub-core/kern/term.c	2021-05-18 14:24:16.920000000 +0200
@@ -69,6 +69,10 @@ grub_putcode_dumb (grub_uint32_t code,
 static void
 grub_xputs_dumb (const char *str)
 {
+  if (!is_dprintf_active)
+    grub_debug_envvar ("text", "%s", str);
+  else
+    grub_debug_envvar ("", "%s", str);
   for (; *str; str++)
     {
       grub_term_output_t term;
diff -urp a/grub-core/normal/term.c b/grub-core/normal/term.c
--- a/grub-core/normal/term.c	2021-05-18 14:25:34.130000000 +0200
+++ b/grub-core/normal/term.c	2021-05-18 14:24:16.960000000 +0200
@@ -413,6 +413,8 @@ read_terminal_list (const char *prefix)
   grub_errno = GRUB_ERR_NONE;
 }
 
+static int string_output_progress = 0;
+
 static void
 putglyph (const struct grub_unicode_glyph *c, struct grub_term_output *term,
 	  int fixed_tab)
@@ -425,6 +427,8 @@ putglyph (const struct grub_unicode_glyp
       .estimated_width = 1
     };
 
+  if (!string_output_progress)
+    grub_debug_envvar ("unicode_char", "%C\n", c->base);
   if (c->base == '\t' && fixed_tab)
     {
       int n;
@@ -581,6 +585,7 @@ print_ucs4_terminal (const grub_uint32_t
       state = &local_state;
     }
 
+  string_output_progress = 1;
   for (i = 0; i < state->bidi_stack_depth; i++)
     putcode_real (state->bidi_stack[i] | (GRUB_UNICODE_LRE & ~0xff),
 		  term, fixed_tab);
@@ -681,6 +686,7 @@ print_ucs4_terminal (const grub_uint32_t
 		    ? ptr + 1 : ptr;
 		  state->backlog_len = last_position - state->backlog_ucs4;
 		  state->backlog_fixed_tab = fixed_tab;
+		  string_output_progress = 0;
 		  return 1;
 		}
 	    }
@@ -746,6 +752,7 @@ print_ucs4_terminal (const grub_uint32_t
       if (fill_right)
 	fill_margin (term, margin_right);
     }
+  string_output_progress = 0;
   return dry_run ? lines : 0;
 }
 
@@ -782,6 +789,7 @@ put_glyphs_terminal (struct grub_unicode
 {
   struct grub_unicode_glyph *visual_ptr;
   int since_last_nl = 1;
+  string_output_progress = 1;
   for (visual_ptr = visual; visual_ptr < visual + visual_len; visual_ptr++)
     {
       if (visual_ptr->base == '\n' && fill_right)
@@ -798,6 +806,7 @@ put_glyphs_terminal (struct grub_unicode
 	      state->backlog_glyphs = visual_ptr + 1;
 	      state->backlog_len = visual_len - (visual_ptr - visual) - 1;
 	      state->backlog_fixed_tab = fixed_tab;
+	      string_output_progress = 0;
 	      return 1;
 	    }
 
@@ -813,6 +822,7 @@ put_glyphs_terminal (struct grub_unicode
   if (fill_right && since_last_nl)
 	fill_margin (term, margin_right);
 
+  string_output_progress = 0;
   return 0;
 }
 
@@ -825,6 +835,7 @@ print_backlog (struct grub_term_output *
   if (!state)
     return 0;
 
+  string_output_progress = 1;
   if (state->backlog_ucs4)
     {
       int ret;
@@ -840,12 +851,14 @@ print_backlog (struct grub_term_output *
 	  state->backlog_len = 0;
 	  state->backlog_ucs4 = 0;
 	}
+      string_output_progress = 0;
       return ret;
     }
 
   if (state->backlog_glyphs)
     {
       int ret;
+      string_output_progress = 1;
       ret = put_glyphs_terminal (state->backlog_glyphs,
 				 state->backlog_len,
 				 margin_left, margin_right, term, state,
@@ -857,9 +870,11 @@ print_backlog (struct grub_term_output *
 	  state->backlog_len = 0;
 	  state->backlog_glyphs = 0;
 	}
+      string_output_progress = 0;
       return ret;
     }
 
+  string_output_progress = 0;
   return 0;
 }
 
@@ -883,6 +898,9 @@ print_ucs4_real (const grub_uint32_t * s
 
   if (!dry_run)
     {
+      if (! is_dprintf_active)
+	grub_debug_envvar ("unicode_text", "%s", grub_ucs4_to_utf8_alloc (str,
+	  last_position - str));
       struct grub_term_coordinate xy;
       if (backlog)
 	state = find_term_state (term);
@@ -1022,6 +1040,8 @@ grub_xnputs (const char *str, grub_size_
 
   if (!unicode_str)
     {
+      if (!is_dprintf_active)
+	grub_debug_envvar ("text", "%s", str);
       for (; msg_len--; str++, msg_len++)
 	{
 	  struct grub_unicode_glyph c =
@@ -1080,6 +1100,8 @@ grub_xnputs (const char *str, grub_size_
 void
 grub_xputs_normal (const char *str)
 {
+  if (is_dprintf_active)
+    grub_debug_envvar ("", "%s", str);
   grub_xnputs (str, grub_strlen (str));
 }
 
diff -urp a/include/grub/misc.h b/include/grub/misc.h
--- a/include/grub/misc.h	2021-03-25 08:39:44.900000000 +0100
+++ b/include/grub/misc.h	2021-05-18 14:31:17.230000000 +0200
@@ -341,6 +341,10 @@ grub_puts (const char *s)
 
 int EXPORT_FUNC(grub_puts_) (const char *s);
 int EXPORT_FUNC(grub_debug_enabled) (const char *condition);
+void EXPORT_FUNC (grub_debug_play_number) (unsigned long x);
+void EXPORT_FUNC (grub_debug_envvar) (const char *condition, const char *fmt, ...)
+     __attribute__ ((format (GNU_PRINTF, 2, 3)));
+extern int EXPORT_VAR (is_dprintf_active);
 void EXPORT_FUNC(grub_real_dprintf) (const char *file,
                                      const int line,
                                      const char *condition,
